rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+//  Or we could instead return a struct with the info
+//       pub struct KeyReturnValue {
+//          key_mode : RsaKeyMode,
+//          modulus: u64,
+//          p: &[u8],
+//          q: &[u8],
+//              ...
+//        }
+
+///
+// Rsa functions. For further documentation please refer to symcrypt.h
+// TODO rest of the documentation
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+/// 2^16 + 1
+const DEFAULT_PUBLIC_EXPONENT: u64 = 0x10001;
+
+/// Rsa Public and Private Key State.
+///
+/// [`RsaKeyPair`] stores a private and public Rsa key.
+pub struct RsaKeyPair {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+}
+
+/// Rsa Public Key State
+///
+/// [`RsaPublicKey`] stores an Rsa public key.
+pub struct RsaPublicKey {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// `RsaKeyUsage` will indicate if the [`RsaKeyPair`] or [`RsaPublicKey`] will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`].
+/// 
+/// [`RsaKeyUsage::Encrypt`] will have both [`RsaKeyUsage::Sign`] and [`RsaKeyUsage::Encrypt`] flags set.
+pub enum RsaKeyUsage {
+    Sign,
+    Encrypt,
+}
+
+// !Review: Should we add the no fips / minimal validation flags? The whole purpose is to eventually get to fips, adding these extra flags might not be needed?
+// #define SYMCRYPT_FLAG_KEY_NO_FIPS               (0x100)
+// #define SYMCRYPT_FLAG_KEY_MINIMAL_VALIDATION    (0x200)
+// #define SYMCRYPT_FLAG_RSAKEY_SIGN       (0x1000) 
+// #define SYMCRYPT_FLAG_RSAKEY_ENCRYPT    (0x2000)
+impl RsaKeyUsage {
+    pub fn to_flag(&self) -> symcrypt_sys::UINT32 {
+        match self {
+            RsaKeyUsage::Sign => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN,
+            RsaKeyUsage::Encrypt => symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_ENCRYPT | symcrypt_sys::SYMCRYPT_FLAG_RSAKEY_SIGN ,
+        }
+    }
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct RsaKeyPairBlob {
+    modulus: Vec<u8>,
+    pub_exp: Vec<u8>,
+    p: Vec<u8>,
+    q: Vec<u8>,
+    d_p: Vec<u8>,
+    d_q: Vec<u8>,
+    crt_coefficient: Vec<u8>,
+    private_exp: Vec<u8>,
+}
+
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct RsaPublicKeyBlob {
+    modulus: Vec<u8>,
+    pub_exp: Vec<u8>,
+}
+
+impl RsaKeyPair {
+    /// `generate_new()` generates a random Rsa key based on the provided parameters.
+    ///
+    /// `n_bits_mod` represents the desired bit length of the Rsa Key.
+    ///
+    /// `pub_exp` takes in an `Option<u64>` that is the public exponent. If `None` is provided, the default `2^16 +1` will be used.
+    ///  
+    /// `key_usage` will indicate if this key will be used for [`RsaKeyUsage::Sign`] or [`RsaKeyUsage::Encrypt`]
+    pub fn generate_new(
+        n_bits_mod: u32,
+        pub_exp: Option<&[u8]>,
```

Comment:
>&[u8] [](http://example.com/codeflow?start=24&length=5)

Why pass 'e' as a byte array? SymCrypt only supports uint64_t anyway.

Reply:
On the wire and in rust, this input will be multi-byte unsigned int, its not normally represented as a `uint64_t`, so we have to do the conversion for the caller.

Reply:
You mean RustTLS represents it as a byte array? I think the conversion should be done in the RustTLS-SymCrypt layer in that case. From a SymCrypt perspective it makes more sense for this to be a u64.